src/rt/sched/behaviourcore.h

```diff
@@ -236,22 +219,23 @@ namespace verona::rt
     Slot* next_slot()
     {
       return (
-        Slot*)((status.load(std::memory_order_acquire) >> NEXT_POINTER_SHIFT) << NEXT_POINTER_SHIFT);
+        Slot*)(status.load(std::memory_order_acquire) & STATUS_NEXT_SLOT_MASK);
     }
 
     /**
-     * Returns true if current slot is a writer or a blocked reader, 
+     * Returns true if current slot is a writer or a blocked reader,
      * otherwise returns false
      */
     bool set_next_slot_reader(Slot* n)
     {
-      assert(((uintptr_t)n & NEXT_POINTER_SHIFT) == 0);
+      // Check that the last two bits are zero
```

Comment:
```C++
    /**
     * Returns true if current slot is a writer or a blocked reader,
     * otherwise returns false
     * Only sets the next value if the active bit has not been set.
     */
    bool set_next_slot_reader(Slot* n)
    {
      // Check that the last two bits are zero
      assert(((uintptr_t)n & ~STATUS_NEXT_SLOT_MASK) == 0);
      assert(next_slot() == nullptr);
      uintptr_t old_status_val = status.load(std::memory_order_acquire);

      uintptr_t new_status_val =
        old_status_val | ((uintptr_t)n) | (STATUS_NEXT_SLOT_READER_FLAG);
      Logging::cout() << "prev slot is_reader" << is_read_only()
                      << " curr reader " << this
                      << "old_status_val: " << old_status_val
                      << " new_status_val: " << new_status_val
                      << Logging::endl;
      // Observed the active bit is set, no need to set next.
      if (is_read_only() && 
            (old_status_val & STATUS_SLOT_ACTIVE_FLAG) ==
              STATUS_SLOT_ACTIVE_FLAG))
        return false;

      // Note that the CAS can only fail if the active bit becomes set.
      return !status.compare_exchange_strong(
        old_status_val, new_status_val, std::memory_order_acq_rel);
    }
```